tools/java/org/unicode/cldr/util/XPathParts.java

```diff
@@ -439,7 +440,19 @@ public String getAttributeValue(int elementIndex, String attribute) {
     }
 
     public void putAttributeValue(int elementIndex, String attribute, String value) {
-        elements.get(elementIndex >= 0 ? elementIndex : elementIndex + size()).putAttribute(attribute, value);
+        elementIndex = elementIndex >= 0 ? elementIndex : elementIndex + size();
+        if (value == null && (elements.get(elementIndex).attributes == null ||
+                              !elements.get(elementIndex).attributes.containsKey(attribute))) {
```

Comment:
It might be slightly more efficient to avoid calling elements.get(elementIndex) repeatedly. Maybe at the beginning: Element e = elements.get(elementIndex), then just use e (except in the last line, where I suppose get() does need to be called again).

Reply:
Agreed, except can do even more.

ea = elements.get(elementIndex).attributes

The last line needs to stay the way it is, because elements.get(elementIndex) can change after make mutable.